AppCenter/AppCenter/Internals/Channel/MSChannelUnitDefault.m

```diff
@@ -414,37 +423,34 @@ - (void)deleteAllLogsWithErrorSync:(NSError *)error {
 #pragma mark - Helper
 
 - (void)enumerateDelegatesForSelector:(SEL)selector withBlock:(void (^)(id<MSChannelDelegate> delegate))block {
-  for (id<MSChannelDelegate> delegate in self.delegates) {
-    if (delegate && [delegate respondsToSelector:selector]) {
-      block(delegate);
+  @synchronized(self.delegates) {
+    for (id<MSChannelDelegate> delegate in self.delegates) {
+      if (delegate && [delegate respondsToSelector:selector]) {
+        block(delegate);
+      }
     }
   }
 }
 
 - (void)notifyFailureBeforeSendingForItem:(id<MSLog>)item withError:(NSError *)error {
-  for (id<MSChannelDelegate> delegate in self.delegates) {
+  @synchronized(self.delegates) {
+    for (id<MSChannelDelegate> delegate in self.delegates) {
 
-    // Call willSendLog before didFailSendingLog
-    if (delegate && [delegate respondsToSelector:@selector(channel:willSendLog:)])
-      [delegate channel:self willSendLog:item];
+      // Call willSendLog before didFailSendingLog
+      if (delegate && [delegate respondsToSelector:@selector(channel:willSendLog:)])
+        [delegate channel:self willSendLog:item];
 
-    // Call didFailSendingLog
-    if (delegate && [delegate respondsToSelector:@selector(channel:didFailSendingLog:withError:)])
-      [delegate channel:self didFailSendingLog:item withError:error];
+      // Call didFailSendingLog
+      if (delegate && [delegate respondsToSelector:@selector(channel:didFailSendingLog:withError:)])
+        [delegate channel:self didFailSendingLog:item withError:error];
+    }
   }
 }
 
 - (void)completedEnqueuingLog:(id<MSLog>)log withInternalId:(NSString *)internalId withSuccess:(BOOL)success {
-  if (success) {
-    [self enumerateDelegatesForSelector:@selector(onFinishedPersistingLog:withInternalId:)
```

Comment:
Yep I had that in mind too, there was no need for 2 different callbacks here 👍 .